index.js

```diff
@@ -202,9 +186,28 @@ Choo.prototype.toString = function (location, state) {
   assert.equal(typeof location, 'string', 'choo.toString: location should be type string')
   assert.equal(typeof this.state, 'object', 'choo.toString: state should be type object')
 
-  this.state.href = location.replace(/\?.+$/, '')
-  this.state.query = nanoquery(location)
-  var html = this.router(location)
+  var href = location.replace(/\?.+$/, '')
+  this._matchRoute(href, location)
+  var html = this._prerender(this.state)
   assert.ok(html, 'choo.toString: no valid value returned for the route ' + location)
   return html.toString()
 }
+
+Choo.prototype._matchRoute = function (location, queryString) {
+  if (location === undefined) location = this._createLocation()
+  if (queryString === undefined) queryString = window.location.search
+  var matched = this.router.match(location)
+  this.state.href = location
+  this.state.query = nanoquery(queryString)
+  this.state.route = matched.route
+  this.state.params = matched.params
+  this.state._handler = matched.handler
+  return this.state
+}
+
+Choo.prototype._prerender = function (state) {
+  var routeTiming = nanotiming("choo.prerender('" + state.route + "')")
+  var res = state._handler(state, this.emitter.emit.bind(this.emitter))
```

Comment:
Should we cache that `this.emitter.emit.bind(this.emitter)` variable somewhere? e.g. in the Choo constructor?

Reply:
Yeah, that'd be great!

Reply:
Directly as `choo.emit`? Then we would have `choo.emitter` and `choo.emit`.